tests/php/test-class-amp-srcset-sanitizer.php

```diff
@@ -0,0 +1,99 @@
+	 */
+	public function data_sanitize() {
+		return [
+			'img_with_valid_srcset'        => [
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     512w, https://example.com/image-2.jpg 1024w   , https://example.com/image-3.jpg 300w, https://example.com/image-4.jpg 768w" width="350" height="150">',
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     512w, https://example.com/image-2.jpg 1024w   , https://example.com/image-3.jpg 300w, https://example.com/image-4.jpg 768w" width="350" height="150">',
+			],
+
+			'img_with_duplicate_img_candidate_but_same_url' => [
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     1024w, https://example.com/image-1.jpg 1024w   , https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg 1x, https://example.com/image-1.jpg 1024w, https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+			],
+
+			'img_with_duplicate_img_candidate_but_different_url' => [
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg, https://example.com/image-1.jpg     1024w, https://example.com/image-2.jpg 1024w   , https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				'<img src="https://example.com/image.jpg" srcset="https://example.com/image.jpg 1x, https://example.com/image-1.jpg 1024w, https://example.com/image-2.jpg 300w, https://example.com/image-3.jpg 768w" width="350" height="150">',
+				[ AMP_Tag_And_Attribute_Sanitizer::DUPLICATE_DIMENSIONS ],
+			],
+
+			'amp_img_srcset_missing_comma' => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="https://example.com/image-1.jpg 1024w https://example.com/image-2.jpg 1024w">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_empty'         => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="">',
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="">',
+				[],
+			],
+
+			'amp_img_srcset_whitespace'    => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="    ">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+
+			'amp_img_srcset_invalid'       => [
+				'<img src="https://example.com/image.jpg" height="100" width="200" srcset="1">',
+				'<img src="https://example.com/image.jpg" height="100" width="200">',
+				[ AMP_Tag_And_Attribute_Sanitizer::INVALID_ATTR_VALUE ],
+			],
+		];
```

Comment:
A few values that are invalid but aren't currently detected by the regex without the change I proposed above:

- `500px`
- `2 x`
- `5kw`

Reply:
@schlessera Like ca3ee36e02c4b42d6ecbf8143cfefedcce2be0dc?